src/AspNetCore/Src/proxymodule.cxx

```diff
@@ -78,56 +78,42 @@ CProxyModule::OnExecuteRequestHandler(
     IHttpEventProvider *
 )
 {
-    HRESULT hr;
-    APPLICATION_MANAGER* pApplicationManager;
-    APPLICATION* pApplication;
-    ASPNETCORE_CONFIG* config;
-    ASPNETCORE_APPLICATION* pAspNetCoreApplication;
-    ASPNETCORE_CONFIG::GetConfig(pHttpContext, &config);
-
-    // TODO store whether we are inproc or outofproc so we don't need to check the config everytime?
-    if (config->QueryIsOutOfProcess())// case insensitive
+    HRESULT hr = S_OK;
+    ASPNETCORE_CONFIG     *pConfig = NULL;
+    APPLICATION_MANAGER   *pApplicationManager = NULL;
+    APPLICATION           *pApplication = NULL;
+    hr = ASPNETCORE_CONFIG::GetConfig(pHttpContext, &pConfig);
+    if(FAILED (hr))
     {
-        m_pHandler = new FORWARDING_HANDLER(pHttpContext);
-        if (m_pHandler == NULL)
-        {
-            hr = E_OUTOFMEMORY;
-            goto Failed;
-        }
-
-        return m_pHandler->OnExecuteRequestHandler();
+        goto Failed;
     }
-    else if (config->QueryIsInProcess())
+
+    pApplicationManager = APPLICATION_MANAGER::GetInstance();
+    if (pApplicationManager == NULL)
     {
-        pApplicationManager = APPLICATION_MANAGER::GetInstance();
-        if (pApplicationManager == NULL)
-        {
-            hr = E_OUTOFMEMORY;
-            goto Failed;
-        }
-
-        hr = pApplicationManager->GetApplication(pHttpContext,
-            &pApplication);
-        if (FAILED(hr))
-        {
-            goto Failed;
-        }
-
-        hr = pApplication->GetAspNetCoreApplication(config, pHttpContext, &pAspNetCoreApplication);
-        if (FAILED(hr))
-        {
-            goto Failed;
-        }
-
-        // Allow reading and writing to simultaneously
-        ((IHttpContext3*)pHttpContext)->EnableFullDuplex();
-
-        // Disable response buffering by default, we'll do a write behind buffering in managed code
-        ((IHttpResponse2*)pHttpContext->GetResponse())->DisableBuffering();
-
-        // TODO: Optimize sync completions
-        return pAspNetCoreApplication->ExecuteRequest(pHttpContext);
+        hr = E_OUTOFMEMORY;
+        goto Failed;
     }
+
+    hr = pApplicationManager->GetApplication(
+                    pHttpContext,
+                    pConfig,
+                    &pApplication);
+    if (FAILED(hr))
+    {
+        goto Failed;
+    }
+
+    m_pHandler = new FORWARDING_HANDLER(pHttpContext, pApplication);
```

Comment:
We are fine to create the application before we create the forwarding handler? I'm not sure if there are any race conditions here.

Reply:
the application is singleton per worker process while the forwarding handler is per request. Inside FORWARDING_HANDLER::ExecuteRequest, the native allocation load the managed one with lock. It ensures only one managed app will be initialized so that there will be no race condition.